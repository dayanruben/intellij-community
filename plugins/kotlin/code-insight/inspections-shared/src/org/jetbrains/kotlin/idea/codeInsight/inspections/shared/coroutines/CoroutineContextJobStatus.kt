// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.analysis.api.components.expressionType
import org.jetbrains.kotlin.analysis.api.components.isSubtypeOf
import org.jetbrains.kotlin.analysis.api.components.resolveToCall
import org.jetbrains.kotlin.analysis.api.resolution.*
import org.jetbrains.kotlin.analysis.api.symbols.KaSymbolModality
import org.jetbrains.kotlin.analysis.api.types.symbol
import org.jetbrains.kotlin.idea.base.analysis.api.utils.hasOrOverridesCallableId
import org.jetbrains.kotlin.psi.KtExpression
import org.jetbrains.kotlin.psi.KtParenthesizedExpression

internal sealed class CoroutineContextJobStatus {

    data object Unknown : CoroutineContextJobStatus()
    data object NoJob : CoroutineContextJobStatus()

    /**
     * Represents a context containing a [kotlinx.coroutines.Job].
     * 
     * Important: The information in this class reflects the most obvious source of a `Job` in the analyzed expression,
     * but does not necessarily represent the **actual** state of the resulting context. 
     * This is because other parts of the context (like unknown `CoroutineContext` parameters) may also contain a `Job`,
     * which could override or affect the values reported here.
     *
     * Example:
     * ```kt
     * fun CoroutineScope.usage(randomContext: CoroutineContext) {
     *   launch(Job() + randomContext) { ... }
     * }
     * ```
     * 
     * When analyzing `Job() + randomContext` expression by [detectFor], the resulting status 
     * will be a [WithJob] with [source] pointing to `Job()` expression, and [isCancellable] set to `false`,
     * even though the `randomContext` may contain 
     * a different [kotlinx.coroutines.Job] element (`NonCancellable`, for example).
     */
    data class WithJob(
        /**
         * The particular expression which introduces a [kotlinx.coroutines.Job] into the context.
         */
        val source: KtExpression,
        /**
         * The cancellability of the [kotlinx.coroutines.Job] present in the context.
         * 
         * The only [kotlinx.coroutines.Job] instance which is considered non-cancellable is exactly 
         * the [kotlinx.coroutines.NonCancellable] object.
         * 
         * It is important to track this property because [kotlinx.coroutines.NonCancellable] is not supposed
         * to be used with just any coroutine builder function.
         * 
         * See documentation on [kotlinx.coroutines.NonCancellable] for more details.
         */
        val isCancellable: Boolean,
    ) : CoroutineContextJobStatus()

    /**
     * Computes the [CoroutineContextJobStatus] after [other] context is being appended to this one with 
     * [kotlin.coroutines.CoroutineContext.plus] operation.
     * 
     * If both this and [other] statuses are [WithJob], then the rightmost [WithJob.source] is used as the resulting one.
     * 
     * See documentation on [WithJob] for the details about combining [WithJob] with [Unknown].
     */
    fun append(other: CoroutineContextJobStatus): CoroutineContextJobStatus {
        return when (other) {
            is Unknown -> this
            is NoJob -> this
            is WithJob -> other
        }
    }

    companion object {
        context(_: KaSession)
        fun detectFor(expression: KtExpression): CoroutineContextJobStatus {
            return expression.detectStatus()
        }

        context(_: KaSession)
        private fun KtExpression?.detectStatus(): CoroutineContextJobStatus {
            val expression = this ?: return Unknown
            
            if (expression is KtParenthesizedExpression) {
                return expression.expression.detectStatus()
            }

            val expressionType = expression.expressionType ?: return Unknown

            val resolvedCall = expression.resolveToCall()?.successfulCallOrNull<KaCallableMemberCall<*, *>>()

            return when {
                expressionType.isSubtypeOf(CoroutinesIds.NonCancellable.ID) -> WithJob(expression, isCancellable = false)
                
                // We assume here that anything that has a `Job` type is a regular cancellable job. Technically, this is not always true,
                // since a random reference to a `Job` can actually point to a `NonCancellable` object instance.
                expressionType.isSubtypeOf(CoroutinesIds.Job.ID) -> WithJob(expression, isCancellable = true)

                expressionType.isSubtypeOf(CoroutinesIds.CoroutineDispatcher.ID) -> NoJob
                
                expressionType.isSubtypeOf(CoroutinesIds.Stdlib.CoroutineContext.ID) && 
                        !expressionType.isSubtypeOf(CoroutinesIds.Job.ID) && 
                        expressionType.symbol?.modality == KaSymbolModality.FINAL -> NoJob

                // TODO: Consider more precise detection for the custom `CoroutineContext`  and `CoroutineContext.Element` implementations
                
                resolvedCall?.symbol?.callableId == CoroutinesIds.currentCoroutineContext -> WithJob(expression, isCancellable = true)
                resolvedCall?.symbol?.callableId == CoroutinesIds.Stdlib.coroutineContext -> WithJob(expression, isCancellable = true)

                resolvedCall?.symbol?.hasOrOverridesCallableId(CoroutinesIds.CoroutineScope.coroutineContext) == true -> WithJob(expression, isCancellable = true)

                resolvedCall is KaFunctionCall && 
                        resolvedCall.symbol.hasOrOverridesCallableId(CoroutinesIds.Stdlib.CoroutineContext.plus) -> handleContextPlusCall(resolvedCall)

                resolvedCall is KaFunctionCall &&
                        resolvedCall.symbol.hasOrOverridesCallableId(CoroutinesIds.Stdlib.CoroutineContext.minusKey) -> handleContextMinusKeyCall(resolvedCall)

                else -> Unknown
            }
        }

        context(_: KaSession)
        private fun handleContextPlusCall(resolvedCall: KaFunctionCall<*>): CoroutineContextJobStatus {
            val dispatchReceiver = resolvedCall.partiallyAppliedSymbol.dispatchReceiver as? KaExplicitReceiverValue

            val leftStatus = dispatchReceiver?.expression.detectStatus()
            val rightStatus = resolvedCall.argumentMapping.keys.singleOrNull().detectStatus()

            return leftStatus.append(rightStatus)
        }

        context(_: KaSession)
        private fun handleContextMinusKeyCall(resolvedCall: KaFunctionCall<*>): CoroutineContextJobStatus {
            val dispatchReceiver = resolvedCall.partiallyAppliedSymbol.dispatchReceiver as? KaExplicitReceiverValue

            val originalStatus = dispatchReceiver?.expression.detectStatus()
            val keyToRemove = resolvedCall.argumentMapping.keys.singleOrNull()

            return if (keyToRemove?.expressionType?.isSubtypeOf(CoroutinesIds.Job.Key.ID) == true) {
                NoJob
            } else {
                originalStatus
            }
        }
    }
}