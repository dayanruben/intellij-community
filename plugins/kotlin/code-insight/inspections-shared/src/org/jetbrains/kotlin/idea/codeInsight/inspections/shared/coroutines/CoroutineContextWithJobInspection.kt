// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import com.intellij.codeInspection.InspectionManager
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.util.TextRange
import com.intellij.psi.SmartPsiElementPointer
import com.intellij.psi.createSmartPointer
import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.analysis.api.resolution.successfulFunctionCallOrNull
import org.jetbrains.kotlin.analysis.api.resolution.symbol
import org.jetbrains.kotlin.idea.base.resources.KotlinBundle
import org.jetbrains.kotlin.idea.codeinsight.api.applicable.inspections.KotlinApplicableInspectionBase
import org.jetbrains.kotlin.name.CallableId
import org.jetbrains.kotlin.psi.KtCallExpression
import org.jetbrains.kotlin.psi.KtExpression
import org.jetbrains.kotlin.psi.KtVisitor
import org.jetbrains.kotlin.psi.callExpressionVisitor

internal class CoroutineContextWithJobInspection : KotlinApplicableInspectionBase<KtCallExpression, CoroutineContextWithJobInspection.Context>() {
    override fun InspectionManager.createProblemDescriptor(
        element: KtCallExpression,
        context: Context,
        rangeInElement: TextRange?,
        onTheFly: Boolean
    ): ProblemDescriptor {
        val description = if (context.jobIsCancellable) {
            KotlinBundle.message("inspection.coroutine.context.with.job.description.cancellable", context.usedBuilder.callableName)
        } else {
            KotlinBundle.message("inspection.coroutine.context.with.job.description.non.cancellable", context.usedBuilder.callableName) 
        }
        
        return createProblemDescriptor(
            /* psiElement = */ context.jobSource.element ?: element,
            /* rangeInElement = */ null,
            /* descriptionTemplate = */ description,
            /* highlightType = */ ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
            /* onTheFly = */ onTheFly,
        )
    }

    override fun buildVisitor(
        holder: ProblemsHolder,
        isOnTheFly: Boolean
    ): KtVisitor<*, *> = callExpressionVisitor {
        visitTargetElement(it, holder, isOnTheFly)
    }

    class Context(val jobSource: SmartPsiElementPointer<KtExpression>, val jobIsCancellable: Boolean, val usedBuilder: CallableId)

    override fun KaSession.prepareContext(element: KtCallExpression): Context? {
        val functionCall = element.resolveToCall()?.successfulFunctionCallOrNull() ?: return null

        val matchedBuilderInfo = SUPPORTED_BUILDERS.find { it.callableId == functionCall.symbol.callableId } ?: return null

        val (coroutineContextArgument, _) = functionCall.argumentMapping.entries
            .find { (_, paramSymbol) -> paramSymbol.returnType.isSubtypeOf(CoroutinesIds.Stdlib.CoroutineContext.ID) }
            ?: return null

        val jobStatus = CoroutineContextJobStatus.detectFor(coroutineContextArgument)

        if (jobStatus !is CoroutineContextJobStatus.WithJob) {
            // we're only interested in the case when there's a job in the context
            return null
        }

        /**
         * We want to warn about detected jobs in two situations:
         * 
         * 1. The job is present and is cancellable. This should be reported for all builders.
         * 2. The job is non-cancellable, but the builder does not support non-cancellable jobs.
         */
        val jobShouldBeReported = jobStatus.isCancellable || !matchedBuilderInfo.allowsNonCancellableJob 

        if (!jobShouldBeReported) return null

        return Context(
            jobSource = jobStatus.source.createSmartPointer(),
            jobIsCancellable = jobStatus.isCancellable,
            usedBuilder = matchedBuilderInfo.callableId,
        )
    }
}

private data class CoroutineBuilderInfo(
    val callableId: CallableId, 
    val allowsNonCancellableJob: Boolean = false,
)

private val SUPPORTED_BUILDERS: Set<CoroutineBuilderInfo> = setOf(
    CoroutineBuilderInfo(CoroutinesIds.launch),
    CoroutineBuilderInfo(CoroutinesIds.async),
    CoroutineBuilderInfo(CoroutinesIds.future),
    CoroutineBuilderInfo(CoroutinesIds.Channels.produce),
    CoroutineBuilderInfo(CoroutinesIds.promise),

    CoroutineBuilderInfo(CoroutinesIds.withContext, allowsNonCancellableJob = true),
    CoroutineBuilderInfo(CoroutinesIds.Flows.flowOn, allowsNonCancellableJob = true),
)
