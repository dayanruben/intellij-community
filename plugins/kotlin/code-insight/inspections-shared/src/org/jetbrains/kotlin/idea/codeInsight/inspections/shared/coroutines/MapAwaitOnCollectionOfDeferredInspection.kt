// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import com.intellij.codeInspection.ProblemsHolder
import com.intellij.codeInspection.util.InspectionMessage
import com.intellij.codeInspection.util.IntentionFamilyName
import com.intellij.modcommand.ModPsiUpdater
import com.intellij.openapi.project.Project
import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.analysis.api.resolution.KaExplicitReceiverValue
import org.jetbrains.kotlin.analysis.api.resolution.successfulFunctionCallOrNull
import org.jetbrains.kotlin.analysis.api.resolution.successfulVariableAccessCall
import org.jetbrains.kotlin.analysis.api.resolution.symbol
import org.jetbrains.kotlin.analysis.api.symbols.KaFunctionSymbol
import org.jetbrains.kotlin.idea.base.psi.imports.addImport
import org.jetbrains.kotlin.idea.base.resources.KotlinBundle
import org.jetbrains.kotlin.idea.codeinsight.api.applicable.inspections.KotlinApplicableInspectionBase
import org.jetbrains.kotlin.idea.codeinsight.api.applicable.inspections.KotlinModCommandQuickFix
import org.jetbrains.kotlin.idea.codeinsight.utils.ConvertLambdaToReferenceUtils.singleStatementOrNull
import org.jetbrains.kotlin.idea.refactoring.singleLambdaArgumentExpression
import org.jetbrains.kotlin.name.*
import org.jetbrains.kotlin.psi.*
import org.jetbrains.kotlin.resolve.ImportPath

internal class MapAwaitOnCollectionOfDeferredInspection : KotlinApplicableInspectionBase.Simple<KtCallExpression, Unit>() {
    override fun getProblemDescription(
        element: KtCallExpression,
        context: Unit
    ): @InspectionMessage String {
        return KotlinBundle.message("inspection.map.await.on.collection.of.deferred.description")
    }

    override fun buildVisitor(
        holder: ProblemsHolder,
        isOnTheFly: Boolean
    ): KtVisitor<*, *> = callExpressionVisitor { callExpression ->
        visitTargetElement(callExpression, holder, isOnTheFly)
    }

    override fun KaSession.prepareContext(element: KtCallExpression): Unit? {
        val lambdaArgument = element.singleLambdaArgumentExpression() ?: return null

        if (!isIterableMapFunctionCall(element)) return null
        if (!isLambdaWithSingleAwaitCallOnSingleParameter(lambdaArgument)) return null

        return Unit
    }

    private fun KaSession.isIterableMapFunctionCall(element: KtCallExpression): Boolean {
        val functionCall = element.resolveToCall()?.successfulFunctionCallOrNull() ?: return false
        val actualReceiverType = functionCall.partiallyAppliedSymbol.extensionReceiver?.type ?: return false

        return isIterableMapFunction(functionCall.symbol) &&
                actualReceiverType.isSubtypeOf(StandardClassIds.Collection)
    }

    private fun KaSession.isIterableMapFunction(symbol: KaFunctionSymbol): Boolean {
        return symbol.callableId == KOTLIN_COLLECTIONS_MAP_ID &&
                symbol.receiverParameter?.returnType?.isSubtypeOf(StandardClassIds.Iterable) == true
    }
    private fun KaSession.isLambdaWithSingleAwaitCallOnSingleParameter(lambdaExpression: KtLambdaExpression): Boolean {
        val singleLambdaParameterSymbol = lambdaExpression.functionLiteral.symbol.valueParameters.singleOrNull() ?: return false
        val singleReturnedExpression = singleReturnedExpression(lambdaExpression) as? KtDotQualifiedExpression ?: return false

        val awaitCall = singleReturnedExpression.resolveToCall()?.successfulFunctionCallOrNull() ?: return false

        val explicitAwaitReceiverValue = awaitCall.partiallyAppliedSymbol.dispatchReceiver as? KaExplicitReceiverValue ?: return false
        val explicitReceiverAccessCall = explicitAwaitReceiverValue.expression.resolveToCall()?.successfulVariableAccessCall() ?: return false

        return awaitCall.symbol.callableId == KOTLINX_COROUTINES_DEFERRED_AWAIT_ID &&
                explicitReceiverAccessCall.symbol == singleLambdaParameterSymbol
    }

    private fun KaSession.singleReturnedExpression(lambdaExpression: KtLambdaExpression): KtExpression? {
        val singleStatement = lambdaExpression.singleStatementOrNull() ?: return null

        return when (singleStatement) {
            is KtReturnExpression if (singleStatement.targetSymbol == lambdaExpression.functionLiteral.symbol) -> singleStatement.returnedExpression
            else -> singleStatement
        }
    }

    override fun createQuickFix(
        element: KtCallExpression,
        context: Unit
    ): KotlinModCommandQuickFix<KtCallExpression> {
        return object : KotlinModCommandQuickFix<KtCallExpression>() {
            override fun getFamilyName(): @IntentionFamilyName String {
                return KotlinBundle.message("inspection.map.await.on.collection.of.deferred.replace.with.awaitAll")
            }

            override fun applyFix(
                project: Project,
                element: KtCallExpression,
                updater: ModPsiUpdater
            ) {
                // TODO Use Import insertion API after KTIJ-28838 is fixed
                val alreadyImportedByStarImport =
                    element.containingKtFile.importDirectives.any {
                        it.importPath == ImportPath(KOTLINX_COROUTINES_PACKAGE, isAllUnder = true)
                    }

                if (!alreadyImportedByStarImport) {
                    element.containingKtFile.addImport(KOTLINX_COROUTINES_AWAIT_ALL_ID.asSingleFqName())
                }

                element.replace(KtPsiFactory(project).createExpression("awaitAll()"))
            }
        }
    }
}

private val KOTLIN_COLLECTIONS_MAP_ID: CallableId = CallableId(FqName("kotlin.collections"), Name.identifier("map"))

private val KOTLINX_COROUTINES_PACKAGE: FqName = FqName("kotlinx.coroutines")

private val KOTLINX_COROUTINES_DEFERRED_AWAIT_ID = CallableId(ClassId(KOTLINX_COROUTINES_PACKAGE, Name.identifier("Deferred")), Name.identifier("await"))

private val KOTLINX_COROUTINES_AWAIT_ALL_ID = CallableId(KOTLINX_COROUTINES_PACKAGE, Name.identifier("awaitAll"))