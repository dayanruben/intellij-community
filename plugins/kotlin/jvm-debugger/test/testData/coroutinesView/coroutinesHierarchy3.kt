package coroutinesView

// ATTACH_LIBRARY: maven(org.jetbrains.kotlinx:kotlinx-coroutines-core:1.10.2)-javaagent

import kotlinx.coroutines.*

fun main() {
    runBlocking(CoroutineName("root")) {
        test()
    }
}

suspend fun test() {
    coroutineScope {
        launch(CoroutineName("launch1")) {
            delay(1)
            println()
            delay(1)
        }
        launch(CoroutineName("launch2")) {
            coroutineScope {
                val a = async(CoroutineName("childAsync1")) {
                    val coroutineVar = 1
                    delay(100)
                    println("coroutineVar=$coroutineVar")
                    6
                }

                val b = async(CoroutineName("childAsync2")) {
                    val coroutineVar = 2
                    delay(100)
                    println("coroutineVar=$coroutineVar")
                    7
                }
            }
        }
        withContext(Dispatchers.Default) {
            coroutineScope {
                val a = async(CoroutineName("async1")) {
                    val coroutineVar = 1
                    delay(100)
                    println("coroutineVar=$coroutineVar") // ‚Üê breakpoint
                    6
                }

                val b = async(CoroutineName("async2")) {
                    val coroutineVar = 2
                    delay(100)
                    println("coroutineVar=$coroutineVar")
                    7
                }
                //Breakpoint!
                println("result: ${a.await() * b.await()}")
            }
        }
    }
}