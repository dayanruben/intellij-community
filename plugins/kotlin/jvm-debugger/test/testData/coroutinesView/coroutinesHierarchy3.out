LineBreakpoint created at coroutinesHierarchy3.kt:53
Run Java
Connected to the target VM
coroutinesHierarchy3.kt:53
==== Hierarchy of coroutines =====
root RUNNING
	launch2 SUSPENDED
		childAsync1 SUSPENDED
		childAsync2 SUSPENDED
	async1 SUSPENDED
	async2 SUSPENDED
Disconnected from the target VM

Process finished with exit code 0
